src/Routing/EdgeDynamicRouting.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge\Routing;
+
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouteCollection;
+
+/**
+ * Dynamically built routes.
+ */
+class EdgeDynamicRouting {
+
+  /**
+   * Build routes dynamically.
+   */
+  public function routes() {
+    $route_collection = new RouteCollection();
+    $route = new Route('/admin/config/apigee-edge/app/{developer_app_by_name}/credentials',
+      [
+        '_controller' => '\Drupal\apigee_edge\Entity\ListBuilder\DeveloperAppListBuilderForDeveloper::appCredentials',
+        '_title' => 'App credentials',
+        'user' => \Drupal::currentUser()->id(),
```

Comment:
1. Use DI.
2. Should not admins also have access to credentials of _all_ users? Since they have administer/bypass permission... 
3. This should also support team apps.

Reply:
This is how it is currently pulling apps for current user. See Drupal\apigee_edge\ParamConverter\DeveloperAppNameConverter

Reply:
I know,I wrote it :) , but you can still use dependency injection here I think, you just need to implement the right interface. Also, probably it would be better introducing a route like the following if you would like to solve the 2nd point, user/[uid]/apps/[app]/api-keys/[api-key] and a shortcut.
For the 3rd point, you need a team app spec endpoint.

Reply:
good point. I made the change and added two routes:
/api-keys/developer/{uid}/{app_name}
/api-keys/team/{team}/{app_name}